### PR TITLE
XS✔ ◾ ✨ New Rule - Do you know the key principles of API Design?

### PR DESCRIPTION
**Pull request by jeoffreyfischer**
>
> 1. What triggered this change? (PBI link, Email Subject, conversation + reason, etc)

✏️ I saw a great video about key principles of API Design and realised we don't have a rule about it

> 2. What was changed?

✏️ I have added a new rule

> 3. Did you do pair or mob programming (list names)?

✏️ No

### CODE REVIEW

**Review by tiagov8 — COMMENTED** [review]
Hi @jeoffreyfischer 

I commited some changes to your new rule, so leaving for someone else to approve ;)

File: rules/key-principles-of-api-design/rule.md [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+
+---
+
+APIs are everywhere in our daily lives—from social media updates and online shopping to weather forecasts and GPS navigation. These interfaces allow different applications to connect seamlessly, providing the backbone for countless modern conveniences.
+
+Building a high-quality API means adhering to best practices that enhance clarity, reliability, and security, creating a more consistent experience for developers and users.
+
+<!--endintro-->
+
+[Video: Good APIs Vs Bad APIs: 7 Tips for API Design](https://www.youtube.com/watch?v=_gQaygjm_hg)
```
Comment (bradystroud) [inline rules/key-principles-of-api-design/rule.md:24]: Why not embed the video?

```suggestion
`youtube: https://youtu.be/_gQaygjm_hg`
**Video: Good APIs Vs Bad APIs: 7 Tips for API Design (6 min)**
```

**Review by bradystroud — APPROVED** [review]

File: rules/key-principles-of-api-design/rule.md [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+For endpoints that return lists, apply pagination to prevent overwhelming the client with too much data.
+
+Use query parameters like ?page= and ?limit= to specify page numbers and size, offering a more manageable data experience while improving performance.  
+
+## Tip \#5 - Use clear query strings for sorting  
+
+When supporting sorting, apply clear query strings (e.g., `?sortBy=name&order=asc`).
+
+Consistent sorting parameters allow developers to retrieve and organize data efficiently and minimize confusion in handling API responses.  
+
```
Comment (danielmackay) [inline rules/key-principles-of-api-design/rule.md:58]: Query params should generally be optional making the API easier to consume.

File: rules/key-principles-of-api-design/rule.md [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+## Tip \#6 - Security should not be an afterthought
+
+Security is paramount in API design. Pass sensitive information, like access tokens, securely via headers (e.g., `Authorization: Bearer {token}`).
+
+Enforce HTTPS for encrypted communication and consider OAuth for user authentication and authorization, protecting against unauthorized access.  
+
```
Comment (danielmackay) [inline rules/key-principles-of-api-design/rule.md:64]: Common security headers that should be implemented:
- Content-Security-Policy (CSP)
- Strict-Transport-Security (HSTS)
- X-Content-Type-Options
- X-Frame-Options

Do we have a rule on these?

File: rules/key-principles-of-api-design/rule.md [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+## Tip \#7 - Keep cross-resource references simple
+
+For APIs that reference multiple resources (e.g., `userId` in a post endpoint), keep relationships simple to prevent over-complicating endpoints.
+
+Provide clear references or IDs rather than nested data whenever possible to keep API responses readable and easy to follow.  
+
```
Comment (danielmackay) [inline rules/key-principles-of-api-design/rule.md:70]: Can you provide an example here?

File: rules/key-principles-of-api-design/rule.md [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+## Extra tip - Rate limiting  
+
+Implement rate limiting to control the number of requests per user within a time frame, protecting the API from abuse and ensuring fair resource usage.
+
+Provide appropriate status codes and messages (e.g., `HTTP 429`) when limits are reached to inform users clearly.  
+
```
Comment (danielmackay) [inline rules/key-principles-of-api-design/rule.md:76]: **Extra tip - Caching**

Add this as an extra section

**Review by danielmackay — CHANGES_REQUESTED** [review]
Quite a few comments

File: rules/key-principles-of-api-design/rule.md [hunk 6]
```diff
@@ -0,0 +1,80 @@
+---
+type: rule
+archivedreason:
```
[the remark this answers is not in the document]
Comment (jeoffreyfischer) [reply rules/key-principles-of-api-design/rule.md]: Done

File: rules/key-principles-of-api-design/rule.md [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+<!--endintro-->
+
+`youtube: https://youtu.be/_gQaygjm_hg`
+**Video: Good APIs Vs Bad APIs: 7 Tips for API Design (6 min)**
+
+## Tip \#1 - Use clear naming  
+
+Choose descriptive, intuitive names for endpoints and parameters to improve readability. Adhere to REST conventions, using nouns and verbs logically (e.g., `/users` for accessing user data, `/users/{id}` for specific user information).
+
+Clear naming simplifies usage and reduces misunderstandings.
```
[the remark this answers is not in the document]
Comment (jeoffreyfischer) [reply rules/key-principles-of-api-design/rule.md]: Done - I've added your comment

File: rules/key-principles-of-api-design/rule.md [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+Introduce versioning from the start (e.g., `/v1/resource`) to maintain backward compatibility when updating the API.
+
+Versioning helps users manage changes without breaking existing implementations, allowing them to adopt new features gradually.  
+
+## Tip \#4 - Add pagination for responses  
+
+For endpoints that return lists, apply pagination to prevent overwhelming the client with too much data.
+
+Use query parameters like ?page= and ?limit= to specify page numbers and size, offering a more manageable data experience while improving performance.  
+
```
[the remark this answers is not in the document]
Comment (jeoffreyfischer) [reply rules/key-principles-of-api-design/rule.md]: Done - I've added your comment

File: rules/key-principles-of-api-design/rule.md (hunk 2 above)
Comment (jeoffreyfischer) [reply rules/key-principles-of-api-design/rule.md]: Done - I've added your comment

File: rules/key-principles-of-api-design/rule.md (hunk 5 above)
Comment (jeoffreyfischer) [reply rules/key-principles-of-api-design/rule.md]: As per our conversation, I will also add an extra section about **Compression**

File: rules/key-principles-of-api-design/rule.md (hunk 3 above)
Comment (jeoffreyfischer) [reply rules/key-principles-of-api-design/rule.md]: Done - this was added

File: rules/key-principles-of-api-design/rule.md (hunk 4 above)
Comment (jeoffreyfischer) [reply rules/key-principles-of-api-design/rule.md]: Done - I've added an example

File: rules/key-principles-of-api-design/rule.md (hunk 5 above)
Comment (jeoffreyfischer) [reply rules/key-principles-of-api-design/rule.md]: Done - I've added both sections

File: rules/key-principles-of-api-design/rule.md [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+For example, if a DELETE request removes a record, re-sending it should not throw errors if the record is already deleted.
+
+This prevents accidental duplicate data processing.  
+
+## Tip \#3 - Implement versioning  
+
+Introduce versioning from the start (e.g., `/v1/resource`) to maintain backward compatibility when updating the API.
+
+Versioning helps users manage changes without breaking existing implementations, allowing them to adopt new features gradually.  
+
```
[the remark this answers is not in the document]
Comment (danielmackay) [reply rules/key-principles-of-api-design/rule.md]: @jeoffreyfischer - did you want to make mention of this in the rule?

File: rules/key-principles-of-rest-api-design/rule.md [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+
+Enforce HTTPS for encrypted communication and consider OAuth for user authentication and authorization, protecting against unauthorized access.  
+
+## Tip \#7 - Keep cross-resource references simple
+
+For APIs that reference multiple resources (e.g., `userId` in a post endpoint), keep relationships simple to prevent over-complicating endpoints.
+
+Provide clear references or IDs rather than nested data whenever possible to keep API responses readable and easy to follow.  
+
+For example, aovoid using messy query patameters such as `api/products?user_id=123&product_id=321`. Instead, use clearly defined endpoints like `api/users/123/products/321`.
```
Comment (danielmackay) [inline rules/key-principles-of-rest-api-design/rule.md:81]:
```suggestion
For example, avoid using messy query patameters such as `api/products?user_id=123&product_id=321`. Instead, use clearly defined endpoints like `api/users/123/products/321`.
```

File: rules/key-principles-of-rest-api-design/rule.md [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+
+Versioning helps users manage changes without breaking existing implementations, allowing them to adopt new features gradually.  
+
+## Tip \#4 - Add pagination for responses  
+
+For endpoints that return lists, apply pagination to prevent overwhelming the client with too much data.
+
+When paging parameters are omitted from the request, the API should apply some sensible defaults (e.g. page 1, 50 records).
+
+Use query parameters like ?page= and ?limit= to specify page numbers and size, offering a more manageable data experience while improving performance.
```
Comment (danielmackay) [inline rules/key-principles-of-rest-api-design/rule.md:54]:
```suggestion
Use query parameters like `?page=` and `?limit=` to specify page numbers and size, offering a more manageable data experience while improving performance.  
```

**Review by danielmackay — CHANGES_REQUESTED** [review]
1 comment not addressed. 2 more minor changes suggested.

**Review by tiagov8 — APPROVED** [review]

File: rules/key-principles-of-api-design/rule.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (jeoffreyfischer) [reply rules/key-principles-of-api-design/rule.md]: Yes we have this rule: https://www.ssw.com.au/rules/do-you-provide-versioning/

[the remark this answers is not in the document]
Comment (jeoffreyfischer) [reply rules/key-principles-of-api-design/rule.md]: Done - added comment + link to rule + related rule

File: rules/key-principles-of-rest-api-design/rule.md [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+## Tip \#5 - Use clear query strings for sorting  
+
+When supporting sorting, apply clear query strings (e.g., `?sortBy=name&order=asc`). Query params should generally be optional making the API easier to consume.
+
+Consistent sorting parameters allow developers to retrieve and organize data efficiently and minimize confusion in handling API responses.  
+
+## Tip \#6 - Security should not be an afterthought
+
+Security is paramount in API design. Pass sensitive information, like access tokens, securely via headers (e.g., `Authorization: Bearer {token}`).  
+
```
Comment (lukecookssw) [inline rules/key-principles-of-rest-api-design/rule.md:76]: I don't like this sentence. It implies that HTTP headers are more "secure" than alternatives (body? query params?).

Headers are just as insecure as anything else. I think a better phrase would be something like:

Security is paramount when building REST APIs. Most REST APIs are hosted online, and you don't want to rely on "security via obscurity". Ensure you spend time hardening your surface area.
Common sense approaches include protecting your endpoints via short-lived access tokens (even for seemingly benign functionality), as well as your typical security headers such as:

**Review by lukecookssw — CHANGES_REQUESTED** [review]
one comment, but aside from that it looks good

**Review by danielmackay — APPROVED** [review]
LGTM

File: rules/key-principles-of-rest-api-design/rule.md (hunk 12 above)
Comment (jeoffreyfischer) [reply rules/key-principles-of-rest-api-design/rule.md]: Done - I've updated the comment

**Review by lukecookssw — APPROVED** [review]
🚀

**Pull request merged**